File: repo/tests/selenium/lib/pages/login_page.rb
# -*- encoding : utf-8 -*-
# Wikidata UI tests
#
# Author:: Tobias Gritschacher (tobias.gritschacher@wikimedia.de)
# License:: GNU GPL v2+
#
# page object for login page

class LoginPage
  include PageObject

  page_url WIKI_REPO_URL + 'index.php?title=Special:UserLogin'
  text_field(:username, :id => 'wpName1')
  text_field(:password, :id => 'wpPassword1')
  button(:login, :id => 'wpLoginAttempt')
  link(:logout, :xpath => "//li[@id='pt-logout']/a")

  def login_with(username, password)
    self.username = username
    self.password = password
    login
  end

  def logout_user
    if logout?
      logout
    end
  end
end

File: cypress/support/pageObjects/LoginPage.ts
import Chainable = Cypress.Chainable;

export class LoginPage {

	private static get LOGIN_FORM_SELECTORS(): Record<string, string> {
		return {
			USERNAME: '#wpName1',
			PASSWORD: '#wpPassword1',
			LOGIN_BUTTON: '#wpLoginAttempt',
		};
	}

	public setUsername( username: string ): Chainable {
		return cy.get( LoginPage.LOGIN_FORM_SELECTORS.USERNAME ).clear().type( username );
	}

	public setPassword( password: string ): Chainable {
		return cy.get( LoginPage.LOGIN_FORM_SELECTORS.PASSWORD ).clear().type( password );
	}

	private getLoginButton(): Chainable {
		return cy.get( LoginPage.LOGIN_FORM_SELECTORS.LOGIN_BUTTON );
	}

	public open(): Chainable {
		return cy.visitTitle( 'Special:UserLogin' );
	}

	public login( username: string, password: string ): Chainable {
		return this.open()
			.then( () => this.setUsername( username ) )
			.then( () => this.setPassword( password ) )
			.then( () => this.getLoginButton().click() );
	}

}
